src/main/java/carpet/script/value/ValueConversions.java

```diff
@@ -499,9 +505,34 @@ public static ItemStack getItemStackFromValue(Value value, boolean withCount, Re
         }
         ItemStack itemInput = NBTSerializableValue.parseItem(name, nbtTag, regs);
         itemInput.setCount(count);
+        System.out.println("itemInput: " + itemInput);
         return itemInput;
     }
 
+    public static Value getComponentsFromItemStack(ItemStack stack, RegistryAccess regs) {
+        var lst = stack.getComponents().keySet().stream().filter(Predicate.not(DataComponentType::isTransient))
+                .collect(Collectors.toMap(
+                        ck -> ValueConversions.of(BuiltInRegistries.DATA_COMPONENT_TYPE.getKey(ck)),
+                        (DataComponentType ck) -> {
+                            var cvalue = stack.get(ck);
+                            return switch (cvalue) {
+                                case Number n -> NumericValue.of(n);
+                                case Boolean b -> BooleanValue.of(b);
+                                default -> {
+                                    Tag res = (Tag) ck.codec()
+                                            .encodeStart(
+                                                    regs.createSerializationContext(NbtOps.INSTANCE),
+                                                    cvalue)
+                                            .result().orElse(null);
+                                    yield NBTSerializableValue.of(res);
+                                }
+                            };
+                        }));
+
+        MapValue mapValue = MapValue.wrap(Map.of(StringValue.of("components"), MapValue.wrap(lst), StringValue.of("id"), StringValue.of(stack.getItem().toString()), StringValue.of("count"), NumericValue.of(stack.getCount())));
```

Comment:
if you pack all of those k-v pairs to a NBT tag again, why you left the switch there?
you can let Number&Boolean use the default case too.


and you no longer need to use a NBTSerializableValue to warp the `res` tag.

Reply:
I've changed that to your implementation, removing the part that took the item from the main hand in case of no values passed to the function. I also removed unedited files. It should be fine now, I guess

Reply:
then you please edit your Overview, Reason and Request for Review......